### PR TITLE
Support Sidekiq 7

### LINKED ISSUE #357: Error with redis 5.x gem
With redis 5.x gem (which use redis-client 0.7.x), Sidekiq cron prevent Sidekiq from starting due to error in redis_client.

```bash
[18] pry(main)> Sidekiq::Cron::Job.load_from_array [{name: 'unique_job', class: 'ReportMailingWorker', queue: 'mailer', cron: '1 0 * * *'}]
TypeError: Unsupported command argument type: NilClass
from /home/user/.asdf/installs/ruby/2.7.6/lib/ruby/gems/2.7.0/gems/redis-client-0.7.1/lib/redis_client/command_builder.rb:75:in `block in generate'
```

With redis 4.8.x (which does not use redis-client), Sidekiq cron loaded successfully.

Currently I version fixed the redis gem to '~> 4.8' and sidekiq cron schedules are loaded successfully.

### LINKED ISSUE #370: Error when running with Sidekiq v7.0.0.beta1
I know it is very early to update to the v7, but I believe that now is the moment for us to start working on it:

Error:

```
/usr/local/bundle/gems/sidekiq-cron-1.7.0/lib/sidekiq/cron/job.rb:16:in `<class:Job>': undefined method `version' for nil:NilClass (NoMethodError)

      REDIS_EXISTS_METHOD = Gem.loaded_specs['redis'].version < Gem::Version.new('4.2') ? :exists : :exists?
                                                     ^^^^^^^^
	from /usr/local/bundle/gems/sidekiq-cron-1.7.0/lib/sidekiq/cron/job.rb:8:in `<module:Cron>'
```

### LINKED ISSUE #378: Sidekiq 7 incompatibility errors
**Gemfile**
```
source 'https://rubygems.org'
ruby '3.1.2'

# Redis
# https://github.com/redis/redis-rb
gem 'redis'

# Async processing
# https://github.com/mperham/sidekiq
gem 'sidekiq'

# https://github.com/ondrejbartas/sidekiq-cron
# Free scheuduled jobs
gem 'sidekiq-cron'
```

**Gemfile.lock**
```
    sidekiq (7.0.1)
      concurrent-ruby (< 2)
      connection_pool (>= 2.3.0)
      rack (>= 2.2.4)
      redis-client (>= 0.9.0)
    sidekiq-cron (1.8.0)
      fugit (~> 1)
      sidekiq (>= 4.2.1)
```

**Error**
```sh
05:20 $  bundle exec sidekiq start
undefined method `options' for Sidekiq:Module

      new_version? ? Sidekiq[key] : Sidekiq.options[key]
                                           ^^^^^^^^
/Users/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sidekiq-cron-1.8.0/lib/sidekiq/options.rb:6:in `[]'
/Users/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sidekiq-cron-1.8.0/lib/sidekiq/cron/schedule_loader.rb:7:in `block in <top (required)>'
/Users/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sidekiq-7.0.1/lib/sidekiq.rb:98:in `configure_server'
/Users/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sidekiq-cron-1.8.0/lib/sidekiq/cron/schedule_loader.rb:6:in `<top (required)>'
/Users/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sidekiq-cron-1.8.0/lib/sidekiq/cron.rb:4:in `require'
/Users/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sidekiq-cron-1.8.0/lib/sidekiq/cron.rb:4:in `<top (required)>'
/Users/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sidekiq-cron-1.8.0/lib/sidekiq-cron.rb:2:in `require'
/Users/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sidekiq-cron-1.8.0/lib/sidekiq-cron.rb:2:in `<top (required)>'
```

### PR DESCRIPTION
**Pull request by serprex**
Sidekiq v7 support 🎉 

Closes #378
Closes #370
Closes #357

### CODE REVIEW

**Comment by markets** [comment]
@serprex is this change enough to make it work with Sidekiq 7?

File: lib/sidekiq/options.rb
Comment (serprex) [inline lib/sidekiq/options.rb:16]: `Gem::Version` being weird about versions with beta/rc may make it preferable to wait until `7.0` is officially released to merge. Wasn't able to figure out a clean way to do this check

**Comment by serprex** [comment]
Unfortunately there was more than merely fixing redis check

With updates Sidekiq no longer crashes at start. Will take some time to validate further

**Comment by serprex** [comment]
So testing locally this is working, but unfortunately I can't test this on prod since Azure's Redis service only supports up to 6.0, while Sidekiq 7 requires at least 6.2

**Comment by markets** [comment]
hey @serprex 👋🏼 Sidekiq 7.0 is out! maybe it's time to update and finalize this PR? There are some failing tests too.

Another thing we need to implement is CI matrix with different Sidekiq versions, so we can automatically be sure changes works in both versions.

**Comment by serprex** [comment]
Are you able to allow CI to run without approval? I'm not sure how to run tests locally

**Comment by markets** [comment]
hey @serprex, you can run the tests locally by (https://github.com/sidekiq-cron/sidekiq-cron#testing):

    $ bundle install
    $ bundle exec rake test

**Comment by serprex** [comment]
I was able to get tests passing with Ruby 3.1 on my system. Needs more work to not break backwards compatibility

File: lib/sidekiq/cron/poller.rb
Comment (serprex) [inline lib/sidekiq/cron/poller.rb:35]: This is wrong, it should pull from the passed in config

File: lib/sidekiq/cron/poller.rb
Comment (serprex) [inline lib/sidekiq/cron/poller.rb:10]: the parameter is a `Sidekiq::Config` in 7. The inherited method's parameter is non-optional, so should probably bring this back, also to keep compatibility with before-6.5

File: test/unit/poller_test.rb
Comment (serprex) [inline test/unit/poller_test.rb:5]: Not sure what this `REDIS` thing is about. Sidekiq 7 doesn't support `Sidekiq.redis = ...`, seems `redis=` method exists on config, seemed most straightforward to gut this constant

File: lib/sidekiq/cron/poller.rb
Comment (serprex) [inline lib/sidekiq/cron/poller.rb:10]: Alternative is to try have default be dynamic between {} & `Sidekiq::Config.new` or `Sidekiq.default_configuration`. But I think it's better to match the underlying method, where this parameter is non-optional. So it's a mistake to call this without a parameter 6.5 & beyond. So maybe an argument error should be raised when config is nil in >=6.5 case

File: lib/sidekiq/cron/job.rb
Comment (haffla) [inline lib/sidekiq/cron/job.rb:560]:
```suggestion
        1 == Sidekiq.redis do |conn|
          conn.public_send(REDIS_EXISTS_METHOD, redis_key(name))
        end
```

Comment (serprex) [reply lib/sidekiq/cron/job.rb:560]: That'll resolve as false when the method returns a bool (`1 == true` is `false`), which it does before Sidekiq 7

Comment (haffla) [reply lib/sidekiq/cron/job.rb:560]: Oooooh sorry then. I didn't know that it could return a bool. What about saving the result in some variable and then `var == true || var == 1`?

**Comment by serprex** [comment]
Due to Sidekiq 7 not supporting Ruby 2.6, there's a funny situation where CI is testing pre-7 sidekiq on ruby 2.6

A testing matrix would be preferable, but for now that exists to validate this PR & I can leave the github actions work to someone else

**Comment by markets** [comment]
Thanks @serprex! 

> Due to Sidekiq 7 not supporting Ruby 2.6, there's a funny situation where CI is testing pre-7 sidekiq on ruby 2.6

Nice! such a lucky context :)) But yeah, ideally we should configure the CI to run tests against different Sidekiq versions (we can propably just use something like [Appraisal](https://github.com/thoughtbot/appraisal)). That can be done in a different PR.

I'm ok ✅ with this branch! I'd like more eyes from maintainers/community on this PR 🤝. Some testing in real apps would be really appreciated (using Sidekiq v6.5 and v7). I'll do it with my team during this week in some of our apps.

You can test this branch by:

```ruby
gem 'sidekiq-cron', github: 'serprex/sidekiq-cron', branch: 'master'
```

File: lib/sidekiq/cron/job.rb
Comment (runephilosof-karnovgroup) [inline lib/sidekiq/cron/job.rb:24]: According to the [documentation](https://github.com/mperham/sidekiq/blob/main/docs/7.0-Upgrade.md#redis-client) `Sidekiq.redis` is now returning a `RedisClient`
`zadd` is not part of the api of [redis-client](https://github.com/redis-rb/redis-client)

Comment (runephilosof-karnovgroup) [reply lib/sidekiq/cron/job.rb:24]: Also https://github.com/mperham/sidekiq/blob/main/docs/capsule.md#redis-pools
> Sidekiq components and add-ons should avoid using Sidekiq.redis or Sidekiq.logger. Instead use the implicit redis or logger methods available on Sidekiq::Component, Sidekiq::Capsule or Sidekiq::{Client,Server}Middleware. Jobs may continue to use Sidekiq.redis, they will automagically use the capsule's pool of concurrency connections.

Comment (serprex) [reply lib/sidekiq/cron/job.rb:24]: `Sidekiq.redis` returns a `RedisClientAdapter` which handles mapping methods in a way that's generally compatible with previous redis library: https://github.com/mperham/sidekiq/blob/main/lib/sidekiq/redis_client_adapter.rb

I'll look into creating a helper method for avoiding `Sidekiq.redis` 7 onward

Comment (serprex) [reply lib/sidekiq/cron/job.rb:24]: Reading over, there'll be more issues than simply updating how redis is invoked. `Sidekiq::Options` as a whole is incorrectly implemented since it goes to `Sidekiq.default_configuration` rather than the context specific config

Leaning towards that conversion being better as a follow up PR, potentially when sidekiq-cron drops support for older versions of sidekiq

**Comment by markets** [comment]
**UPDATE** 
- Just tested this branch in a Rails app with Sidekiq v6.5 and it seems to work fine!
- #378 Here some people, with Sidekiq v7, report that it works fine too!

I'll give some more days for real testing... if nobody reports new issues, I'll merge and release.

@ondrejbartas @honzasterba are you ok with that plan?

**Comment by honzasterba** [comment]
Awesome! Great work!

**Review by honzasterba — DISMISSED** [review]

File: test/unit/launcher_test.rb
Comment (markets) [inline test/unit/launcher_test.rb:14]:
```suggestion
      Sidekiq::Launcher.new(Sidekiq.respond_to?(:default_configuration) ? Sidekiq.default_configuration : Sidekiq)
```

Comment (markets) [reply test/unit/launcher_test.rb:14]: @serprex ☝🏼 just reverting this extra whitespace

File: lib/sidekiq/cron/job.rb
Comment (markets) [inline lib/sidekiq/cron/job.rb:413]: @serprex maybe this change closes #354 too?

Comment (serprex) [reply lib/sidekiq/cron/job.rb:413]: I don't believe so, previous redis lib would stringify parameters, while new one choked. But with timestamp it should use iso8601 to stringify. I think the other PR should be merged

**Comment by markets** [comment]
Hi @serprex did you know why CI is crashing after last run? That change is totally unrelated, so maybe some deps update?

**Comment by serprex** [comment]
I don't. I won't have a dev environment until next week

**Comment by araluce** [comment]
@markets @serprex probably related with https://github.com/mperham/sidekiq/pull/5621

File: lib/sidekiq/cron/job.rb
Comment (mobilutz) [inline lib/sidekiq/cron/job.rb:403]:
```suggestion
          klass: @klass.to_s,
```
It seems the indentation broke.

**Comment by markets** [comment]
@serprex There are still 4 failing tests with:

    NoMethodError: undefined method `starts_with?'

Should we wait for next Sidekiq release? https://github.com/mperham/sidekiq/compare/v7.0.1...main

**Comment by serprex** [comment]
Yes, 7.0.2 will fix issue, not sure when he intends to release that tho

**Comment by loust333** [comment]
@markets I tested in our app in development:

"sidekiq-cron", github: "serprex/sidekiq-cron", branch: "master"

Everything was fine 😃 

I'm going to test in the staging environment.

**Comment by tisba** [comment]
Looks like Sidekiq 7.0.2 could be [released this week](https://github.com/mperham/sidekiq/pull/5621#issuecomment-1329546850) 🎉

**Comment by tisba** [comment]
And it's out: https://rubygems.org/gems/sidekiq/versions/7.0.2 🥳

**Comment by loust333** [comment]
Deployed in production sidekiq 7.0.2 with this pull request and works like a charm! 🎉

**Comment by markets** [comment]
Nice! and CI is also green again 🟢

I'll cut a new release during the day! Thank you everybody for helping on this one, specially to @serprex 🤘🏼

**Review by markets — APPROVED** [review]

**Comment by markets** [comment]
New release `v1.9.0` published to RubyGems 🚀 Thank you everybody again! 

Let's go for the next one 💪🏼

**Pull request merged**
